lib/stream.fz

```diff
@@ -243,3 +243,41 @@ stream(redef T type) ref : Sequence<T> is
     while hasNext
     else
       r
+
+
+# streams -- unit type defining features related
+# to streams but not requiring an instance
+#
+streams is
+
+  # creates a - potentially infinite - stream by using given supplier
+  # end of stream is reached once supplier returns nil
+  # note that the supplier is called only once for each hasNext/next cycle
+  #
+  generate(T type, supplier () -> option T) stream T is
+
+    not_initialized is
+    depleted is
+    val := mut (choice T not_initialized depleted) not_initialized
+
+    read =>
+      match val.get
+        not_initialized not_initialized =>
```

Comment:
I still find this kind of match-cases confusing. Better use `_`
``` 
  _ not_initialized =>
``` 
or do not declare a field at all
``` 
  not_initialized =>
```